GSA-TTS/FAC
# Cross validation and waivers for prior references

Closes https://github.com/GSA-TTS/FAC/issues/4444

In this PR:
- Verifying existence of prior findings reference number when listed in a finding. Here are the steps for doing this:
  - Waiver exists -> Valid.
  - No prior references in the report -> Valid.
  - The report is from before 2023 (because UEIs only consistent starting 2022) -> Valid.
  - Fetch all disseminated reports for the UEI and look for the prior reference in those. Not found -> Invalid.
- Able to create waivers in the admin panel
- Unit tests
- Tweaking Cypress tests and workbooks to use 2023 submission dates. This is because reports earlier than this skip prior reference validation.

<img width="573" alt="image" src="https://github.com/user-attachments/assets/bd1dbdca-3e6f-4fcd-9585-3427fee4db69">

Testing:
- Note: This has a migration, so be sure those run
1. Modify `full-submission.cy.js` so that it only runs one of the three tests, just to make things faster
2. Run the full-submission test and it should pass as normal. This will also create a submission that we will make use of in later tests.
3. Open up `backend/cypress/fixtures/test_workbooks/federal-awards-audit-findings-workbook.xlsx` and go to the Form tab. Set J2 to 'Y' and set K2 to '2023-999'. This indicates that the finding has a prior reference.
4. In `full-submission.js`, comment out everything after and including line 96. This way it will set up a new audit and we can just click the `Begin Validation` button for the remainder of the tests. Run full-submission again to get up to this point. Note the `report_id` for later steps.
5. ❌ Clicking the button now will cause cross validation to fail with an error saying the prior finding couldn't be found. This is correct because it we didn't set up the prior reference number on the report we made in step 2.
6. Go to http://localhost:8000/admin/audit/sacvalidationwaiver/add/ and create a waiver for the `report_id` from step 4, and select "Report has invalid prior reference numbers"
7. ✅ Clicking the button now should pass cross validation since the waiver lets it ignore errors
8. ❌ Delete the waiver you just created so that we can continue to set up a legit submission. Clicking the button should now fail cross validation again.
9. Go to `dissemination_finding`. Find the matching `report_id` from step 4 and modify the `reference_number` to be '2023-999'. This matches what we set up in the workbook in step 3.
10. ✅ Clicking the button now should pass cross validation

## PR Checklist: Submitter


## PR Checklist: Reviewer


The larger the PR, the stricter we should be about these points.

## Pre Merge Checklist: Merger
